### PR TITLE
6X: Don't append duplicated initplans even if referred at multi places

### PR DESCRIPTION
**Pull request by adam8157**
We do that for normal subplans, but it's not needed for initplans which
only executed once.

## Here are some reminders before you submit the pull request
- [x] Add tests for the change

### CODE REVIEW

File: src/test/regress/expected/subselect_gp.out [hunk 1]
```diff
@@ -2193,3 +2193,30 @@ select * from simplify_sub t1 where not exists (select sum(t2.i) from simplify_s
...
+   c_var timestamp;
+BEGIN
+   BEGIN
+      SELECT c INTO c_var FROM lookup_table WHERE a = a_in;
+   END;
+   RETURN c_var;
+END;
+$$
+LANGUAGE plpgsql NO SQL;
+SELECT * FROM foo f WHERE EXISTS (SELECT 1 FROM foo f1 WHERE f.b=f1.b AND f1.c > (SELECT my_lookup('hello')));
```
Comment (kainwen) [inline src/test/regress/expected/subselect_gp.out:2218]: Can we add some data, and to verify the SQL's result is correct.

Thanks.

**Review by kainwen — APPROVED** [review]

**Comment by kainwen** [comment]
Also mention in the title and pr message, this is 6X only.

File: src/test/regress/sql/subselect_gp.sql [hunk 2]
```diff
@@ -892,3 +892,32 @@ select * from simplify_sub t1 where not exists (select sum(t2.i) from simplify_s
...
+BEGIN
+   BEGIN
+      SELECT c INTO c_var FROM lookup_table WHERE a = a_in;
+   END;
+   RETURN c_var;
+END;
+$$
+LANGUAGE plpgsql NO SQL;
+
+SELECT a, b FROM foo f WHERE EXISTS (SELECT 1 FROM foo f1 WHERE f.b=f1.b AND f1.c > (SELECT my_lookup('a')));
```
Comment (asimrp) [inline src/test/regress/sql/subselect_gp.sql:923]: To understand this patch, I tried running this query on 6X.  Without your patch it fails during execution in `ExecInitSubPlan()`.  The problem, however, is not clear to me.  

On PostgreSQL 13, we get one InitPlan and one SubPlan:
```SQL
explain (verbose, costs off) 
SELECT a, b FROM foo f WHERE EXISTS
    (SELECT 1 FROM foo f1 WHERE f.b=f1.b AND f1.c > (SELECT my_lookup('a')));

                             QUERY PLAN                             
--------------------------------------------------------------------
 Append
   ->  Seq Scan on public.foo_a f_1
         Output: f_1.a, f_1.b
         Filter: (SubPlan 2)
         SubPlan 2
           ->  Append
                 InitPlan 1 (returns $0)
                   ->  Result
                         Output: my_lookup('a'::text)
                 ->  Seq Scan on public.foo_a f1_1
                       Filter: ((f1_1.c > $0) AND (f_1.b = f1_1.b))
                 ->  Seq Scan on public.foo_b f1_2
                       Filter: ((f1_2.c > $0) AND (f_1.b = f1_2.b))
   ->  Seq Scan on public.foo_b f_2
         Output: f_2.a, f_2.b
         Filter: (SubPlan 2)
(16 rows)
```

**Review by asimrp — COMMENTED** [review]
The optimizer specific answer file is needed to make PR pipeline green.

Why is this not a problem in master and in upstream PostgreSQL?  Of course, the code is different, I don't see `fixup_subplan_walker()` in master.  Is there a better answer?

File: src/test/regress/sql/subselect_gp.sql [hunk 3]
```diff
@@ -892,3 +892,32 @@ select * from simplify_sub t1 where not exists (select sum(t2.i) from simplify_s
...
+DECLARE
+   c_var timestamp;
+BEGIN
+   BEGIN
+      SELECT c INTO c_var FROM lookup_table WHERE a = a_in;
+   END;
+   RETURN c_var;
+END;
+$$
+LANGUAGE plpgsql NO SQL;
```
[the remark this answers is not in the document]
Comment (adam8157) [reply src/test/regress/sql/subselect_gp.sql:921]: `PRODATAACCESS_NONE`, NO SQL indicates that the routine contains no SQL statements, I guess it is so for PL/pgSQL.

File: src/test/regress/sql/subselect_gp.sql (hunk 2 above)
Comment (adam8157) [reply src/test/regress/sql/subselect_gp.sql:923]: Our understanding:

It was trying to duplicate the `InitPlan 1` for `Seq Scan on public.foo_a f1_1` and `Seq Scan on public.foo_a f1_1`, but that was actually not needed, there was only one `InitPlan 1`, and it was enough.

It is usually needed if it's a normal SubPlan (InitPlan is a special kind of SubPlan).

Comment (asimrp) [reply src/test/regress/sql/subselect_gp.sql:923]: In the PostgreSQL explain plan, I see that SubPlan 2 (and therefore InitPlan 1) are referenced by both the scan nodes (Seq Scan on foo_a and Seq Scan on foo_b).  Does that mean they are duplicated?

Why is Greenplum 7 immune to this problem?

Comment (adam8157) [reply src/test/regress/sql/subselect_gp.sql:923]: SubPlans are not needed to be duplicated on Greenplum 7, even they are referred at multi places. Not quite sure how vanilla PostgreSQL handles that.

```
commit 50239de1810692256fe2c9aaf6474f082668576f
Author: Heikki Linnakangas <hlinnakangas@pivotal.io>
Date:   Tue Sep 24 13:45:50 2019 +0300

    Allow multiple SubPlan references to a single subplan.
```

Comment (asimrp) [reply src/test/regress/sql/subselect_gp.sql:923]: Thanks for that pointer.

**Review by xiong-gang — APPROVED** [review]
LGTM

**Pull request merged**